src/main/java/one/util/streamex/PrefixOps.java

```diff
@@ -299,4 +299,224 @@ public void accept(T next) {
+        private final LongBinaryOperator op;
+        private boolean started;
+        private MyAtomicLong accRef;
+        private long acc;
+        
+        OfUnordLong(Spliterator.OfLong source, LongBinaryOperator op) {
+            super(source);
+            this.op = op;
+        }
+        
+        private static final class MyAtomicLong extends AtomicLong {
+            private boolean init;
+    
+            /**
+             * On the very first call sets the value to {@code x}
+             * @param x the initial value
+             * @return {@code true} if it was the very first call
+             */
+            public synchronized boolean initialize(long x) {
+                if (!init) {
+                    init = true;
+                    set(x);
+                    return true;
+                }
+                return false;
+            }
+        }
+        
+        @Override
+        public Spliterator.OfLong trySplit() {
+            if (started) {
+                return null;
+            }
+            Spliterator.OfLong prefix = source.trySplit();
+            if (prefix == null) {
+                return null;
+            }
+            if (accRef == null) {
+                accRef = new MyAtomicLong();
+            }
+            OfUnordLong pref = (OfUnordLong) doClone();
+            pref.source = prefix;
+            return pref;
+        }
+        
+        @Override
+        public boolean tryAdvance(LongConsumer action) {
+            if (!source.tryAdvance(this)) {
+                return false;
+            }
+            action.accept(acc);
+            return true;
+        }
+        
+        @Override
+        public void forEachRemaining(LongConsumer action) {
+            if (accRef == null) {
+                source.forEachRemaining((LongConsumer) next -> {
+                    if (started) {
+                        acc = op.applyAsLong(acc, next);
+                    } else {
+                        acc = next;
+                        started = true;
+                    }
+                    action.accept(acc);
+                });
+            } else {
+                long[] buf = new long[BUF_SIZE];
+                source.forEachRemaining((LongConsumer) next -> {
+                    if (idx == 0) {
+                        buf[idx++] = next;
+                    } else {
+                        long prev = buf[idx - 1];
+                        buf[idx++] = op.applyAsLong(prev, next);
+                        if (idx == buf.length) {
+                            drain(action, buf);
+                            idx = 0;
+                        }
+                    }
+                });
+                if (idx > 0)
+                    drain(action, buf);
+            }
+        }
+        
+        private void drain(LongConsumer action, long[] buf) {
+            long last = buf[idx - 1];
+            boolean accRefJustInitialized = accRef.initialize(last);

```

Comment:
> M-m-m... I don't get this... AtomicInteger update inside the synchronized block happens exactly once per stream lifecycle

No it's not. The initialize synchronized method is called on the every drain that happens every time the buffer is full and with the same frequency as getAndAccumulate is called. It would probably easier to use `AtomicReference<AtomicLong>` and initialize AtomicReference in a CAS loop to avoid synchronization...

Reply:
> It would probably easier to use AtomicReference<AtomicLong> and initialize AtomicReference in a CAS loop to avoid synchronization...

Tried this, but performance degraded (from 3938,292 ± 106,900 down to 3373,426 ± 261,430  ops/s). Seems that wraping-unwraping longs to-from AtomicLong has significant overhead (or maybe my implementation of this idea was not good enough). Check it out:

```java
static final class OfUnordLong extends PrefixOps<Long, Spliterator.OfLong> implements LongConsumer, Spliterator.OfLong {
        private static final AtomicLong NONE = new AtomicLong();
        private final LongBinaryOperator localOp;
        private final BinaryOperator<AtomicLong> op;
        private boolean started;
        private AtomicReference<AtomicLong> accRef;
        private long acc;
        private LongPrefixBuffer buffer;
        
        OfUnordLong(Spliterator.OfLong source, LongBinaryOperator op) {
            super(source);
            this.localOp = op;
            //noinspection NumberEquality
            this.op = (a, b) -> a == NONE ? b : new AtomicLong(op.applyAsLong(a.get(), b.get()));
        }
    
        private final class LongPrefixBuffer extends PrefixBuffer implements LongConsumer {
            private final long[] buf = new long[BUF_SIZE];
            private long prevBufferLast;
        
            boolean init(Spliterator.OfLong source) {
                if (idx == 0) {
                    int i = 0;
                    while (i < BUF_SIZE && source.tryAdvance(this)) {
                        i++;
                    }
                    if (idx == 0) {
                        return false;
                    }
                
                    AtomicLong last = new AtomicLong(buf[idx - 1]);
                    isFirst = isFirst || accRef.compareAndSet(NONE, last);
                    if (!isFirst) {
                        prevBufferLast = accRef.getAndAccumulate(last, op).get();
                    }
                }
                return true;
            }
        
            void drainOne(LongConsumer action) {
                long value = buf[--idx];
                if (isFirst) {
                    action.accept(value);
                    if (idx == 0) {
                        isFirst = false;
                    }
                } else {
                    action.accept(localOp.applyAsLong(value, prevBufferLast));
                }
            }
        
            void drainAll(LongConsumer action) {
                if (!isInit()) return;
            
                if (isFirst) {
                    for (int i = 0; i < idx; i++) {
                        action.accept(buf[i]);
                    }
                    isFirst = false;
                } else {
                    for (int i = 0; i < idx; i++) {
                        action.accept(localOp.applyAsLong(buf[i], prevBufferLast));
                    }
                }
                idx = 0;
            }
        
            @Override
            public void accept(long value) {
                if (idx == 0) {
                    buf[idx++] = value;
                } else {
                    long prev = buf[idx - 1];
                    buf[idx++] = localOp.applyAsLong(prev, value);
                }
            }
        }
        
        @Override
        public Spliterator.OfLong trySplit() {
            if (started) {
                return null;
            }
            Spliterator.OfLong prefix = source.trySplit();
            if (prefix == null) {
                return null;
            }
            if (accRef == null) {
                accRef = new AtomicReference<>(NONE);
                buffer = new LongPrefixBuffer();
            }
            OfUnordLong pref = (OfUnordLong) doClone();
            pref.source = prefix;
            pref.buffer = new LongPrefixBuffer();
            return pref;
        }
        
        @Override
        public boolean tryAdvance(LongConsumer action) {
            if (accRef == null) {
                if (!source.tryAdvance(this)) {
                    return false;
                }
                action.accept(acc);
            } else {
                if (!buffer.init(source)) {
                    return false;
                }
                buffer.drainOne(action);
            }
            return true;
        }
        
        @Override
        public void forEachRemaining(LongConsumer action) {
            if (accRef == null) {
                source.forEachRemaining((LongConsumer) next -> {
                    if (started) {
                        acc = localOp.applyAsLong(acc, next);
                    } else {
                        acc = next;
                        started = true;
                    }
                    action.accept(acc);
                });
            } else {
                buffer.drainAll(action);
                long[] buf = new long[BUF_SIZE];
                source.forEachRemaining((LongConsumer) next -> {
                    if (idx == 0) {
                        buf[idx++] = next;
                    } else {
                        long prev = buf[idx - 1];
                        buf[idx++] = localOp.applyAsLong(prev, next);
                        if (idx == buf.length) {
                            drain(action, buf);
                            idx = 0;
                        }
                    }
                });
                if (idx > 0)
                    drain(action, buf);
            }
        }
        
        private void drain(LongConsumer action, long[] buf) {
            long last = buf[idx - 1];
            AtomicLong acc = accRef.getAndAccumulate(new AtomicLong(last), op);
            //noinspection NumberEquality
            if (acc != NONE) {
                for (int i = 0; i < idx; i++) {
                    action.accept(localOp.applyAsLong(buf[i], acc.get()));
                }
            } else {
                for (int i = 0; i < idx; i++) {
                    action.accept(buf[i]);
                }
            }
        }
        
        @Override
        public void accept(long next) {
            if (accRef == null) {
                if (started) {
                    acc = localOp.applyAsLong(acc, next);
                } else {
                    started = true;
                    acc = next;
                }
            } else {
                acc = accRef.accumulateAndGet(new AtomicLong(next), op).get();
            }
        }
    }
```

Also, tried to make use of StampedLock to avoid synchronized methods in MyAtomicLong but got no significant performance change, while code became much uglier:

```java
private static final class MyAtomicLongWithStampedLockInit extends AtomicLong {
        private final StampedLock sl = new StampedLock();
        private boolean init;
        
        public boolean initialize(long x) {
            if (!this.init) {
                boolean result = false;
                long stamp = sl.writeLock();
                if (!this.init) {
                    this.init = true;
                    set(x);
                    result = true;
                }
                sl.unlockWrite(stamp);
                return result;
            }
            if (sl.isWriteLocked()) {
                sl.unlockRead(sl.readLock());
            }
            return false;
        }
}
```
	
>The initialize synchronized method is called on the every drain that happens every time the buffer is full

Yes, it's true. My point was that the only operation on AtomicInteger inside this synchronized method happens exactly once per stream lifecycle. After that synchronized method become effectively equivalent to:

```java
public synchronized boolean initialize(long x) {
    if (!true) {}
    return false;
}
```

So it's still a bottleneck, but it seems it's not that narrow after all.

Reply:
Well, my idea about double wrapping was slightly different. Like this (I have an unfinished draft):

```java
AtomicReference<AtomicLong> ref;
// in trySplit:
        if (ref == null) {
            ref = new AtomicReference<>();
        }

        private void drain(LongConsumer action, long[] buf) {
            long last = buf[idx - 1];
            AtomicLong sharedAcc = sharedAcc(last);
            if (sharedAcc == null) {
                for (int i = 0; i < idx; i++) {
                    action.accept(buf[i]);
                }
            }
            long acc = sharedAcc.getAndAccumulate(last, op); // op is just original op
            for (int i = 0; i < idx; i++) {
                action.accept(op.apply(buf[i], acc));
            }
        }

        private AtomicLong sharedAcc(long last) {
            AtomicLong sharedAcc = ref.get();
            if (sharedAcc == null && !ref.compareAndSet(null, new AtomicLong(last))) {
                sharedAcc = ref.get();
            }
            return sharedAcc;
        }
```

I haven't benchmarked it though. Probably it's no better.

Reply:
Wow, nice trick! It definetely better than my implementation of this idea, but unfortunately comparing to current implementation performance slightly degraded.

This is an isolated benchmark for different initialization strategies:

```java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.locks.StampedLock;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

public class InitializeBenchmark {
    private static final class StampedLockInit extends AtomicLong {
        private final StampedLock sl = new StampedLock();
        private boolean init;
        
        public boolean initialize(long x) {
            if (!this.init) {
                boolean result = false;
                long stamp = sl.writeLock();
                if (!this.init) {
                    this.init = true;
                    set(x);
                    result = true;
                }
                sl.unlockWrite(stamp);
                return result;
            }
            if (sl.isWriteLocked()) {
                sl.unlockRead(sl.readLock());
            }
            return false;
        }
    }
    
    private static final class SynchronizedInit extends AtomicLong {
        private boolean init;
        
        /**
         * On the very first call sets the value to {@code x}
         *
         * @param x the initial value
         * @return {@code true} if it was the very first call
         */
        public synchronized boolean initialize(long x) {
            if (!init) {
                init = true;
                set(x);
                return true;
            }
            return false;
        }
    }
    
    private static final class AtomicReferenceInit {
        private final AtomicReference<AtomicLong> ref = new AtomicReference<>();
        /**
         * On the very first call sets the value to {@code x}
         *
         * @param x the initial value
         * @return {@code null} if it was the very first call
         */
        public AtomicLong getOrInitialize(long x) {
            AtomicLong sharedAcc = ref.get();
            if (sharedAcc == null && !ref.compareAndSet(null, new AtomicLong(x))) {
                sharedAcc = ref.get();
            }
            return sharedAcc;
        }
    }
    
    @State(Scope.Benchmark)
    public static class MyState {
        public ForkJoinPool fjp = new ForkJoinPool();
        private AtomicReferenceInit atomicReferenceInit;
        private StampedLockInit stampedLockInit;
        private SynchronizedInit synchronizedInit;
        private List<Callable<Boolean>> atomicReferenceInitTasks;
        private List<Callable<Boolean>> stampedLockInitTasks;
        private List<Callable<Boolean>> synchronizedInitTasks;
        
        @Param({"100000"})
        private int N;
    
        @Setup(Level.Trial)
        public void doSetup() {
            atomicReferenceInit = new AtomicReferenceInit();
            stampedLockInit = new StampedLockInit();
            synchronizedInit = new SynchronizedInit();
            atomicReferenceInitTasks = new ArrayList<>(N);
            stampedLockInitTasks = new ArrayList<>(N);
            synchronizedInitTasks = new ArrayList<>(N);
            for (int i = 0; i < N; i++) {
                long x = ThreadLocalRandom.current().nextLong();
                atomicReferenceInitTasks.add(() -> atomicReferenceInit.getOrInitialize(x) == null);
                stampedLockInitTasks.add(() -> stampedLockInit.initialize(x));
                synchronizedInitTasks.add(() -> synchronizedInit.initialize(x));
            }
        }
    }
    
    @Benchmark
    public void atomicReferenceInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.atomicReferenceInitTasks));
    }
    
    @Benchmark
    public void stampedLockInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.stampedLockInitTasks));
    }
    
    @Benchmark
    public void synchronizedInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.synchronizedInitTasks));
    }
    
    public static void main(String[] args) throws RunnerException {
        Options opt = new OptionsBuilder()
                .include(InitializeBenchmark.class.getSimpleName())
                .build();
        
        new Runner(opt).run();
    }
}
```


Got the following results on my computer:

```
Benchmark                                   (N)   Mode  Cnt   Score   Error  Units
InitializeBenchmark.atomicReferenceInit  100000  thrpt   25  47,046 ± 6,739  ops/s
InitializeBenchmark.stampedLockInit      100000  thrpt   25  48,853 ± 7,217  ops/s
InitializeBenchmark.synchronizedInit     100000  thrpt   25  57,980 ± 0,754  ops/s
```